text/0000-promote-riscv64gc-unknown-linux-gnu-to-tier1-without-host-tools.md

```diff
@@ -0,0 +1,185 @@
+   test binaries for the target on systems they control, or provide such
+   binaries for others to run. (Doing so may require enabling some appropriate
+   "developer mode" on such systems, but must not require the payment of any
+   additional fee or other consideration, or agreement to any onerous legal
+   agreements.)
+
+No hard requirement of signing, verifying, or "approving" binaries exists for the `riscv64gc-unknown-linux-gnu` platform.
+
+> 2.a. The long term viability of the existence of a target specific ecosystem should be clear.
+
+RISC-V has a roughly 9 year history and there are a variety of vendors providing silicon using this instruction set. They include (but are not limited to) [Alibaba Cloud](https://www.alibabagroup.com/), [AllWinner](http://www.allwinnertech.com/), [antimicro](http://antmicro.com/), [BeagleBoard](https://beagleboard.org/), [Deep Computing](https://deepcomputing.io/), [Microchip](https://www.microchip.com/), [RIOS](http://rioslab.org/), [SiFive](https://sifive.com/), [SOPHGO](https://en.sophgo.com/site/index.html), and [StarFive](https://starfivetech.com/).
+
+There is already an existing ecosystem of downstream users of this target. [Debian](https://wiki.debian.org/Ports/riscv64), [Ubuntu](https://ubuntu.com/download/risc-v) and [OpenSUSE](https://en.opensuse.org/openSUSE:RISC-V) all provide `riscv64` distributions of Linux and also package Rust. [Scaleway](https://labs.scaleway.com/en/em-rv1/) is offering `riscv64gc-unknown-linux-gnu` cloud instances.
+
+Some of the ongoing development of this target has been supported by the [RISE Project](https://riseproject.dev/) which represents a broad array of industrial interests including, for example, Google, Intel, NVIDIA, and SiFive.
+
+It is fair to say that the target specific ecosystem has been a viable target for some time now, and that this is likely to continue into the long term.
+
+> 2.b. The long term viability of supporting the target should be clear.
+
+It is hard to concretely quantify this aspect. This work was initiated and supported by the [RISE Project](https://riseproject.dev/) and there is an intention to continue to support the target and eventually propose a *Tier 1 with Host Tools* RFC when sufficiently fast hardware exists.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Adopting the platform would require additional commitments by the Rust project. Future contributions may impact the target and cause changes to become delayed or halted entirely due to problems on the target.
+
+In general, it should be uncomplicated for contributors to build for and use a `riscv64gc-unknown-linux-gnu` emulator like `qemu`, `docker`, or `lima`. Additionally, the platform is a `*-unknown-linux-gnu` target which is generally quite well understood, contributors do not need to learn what could be an otherwise unfamiliar operating system.
+
+This target does not place significant burdens on the project that would not be present on any other target.
```

Comment:
> In general, it should be uncomplicated for contributors to build for and use a `riscv64gc-unknown-linux-gnu` emulator like `qemu`, `docker`, or `lima`.

+1 to what Jubilee said. If as a compiler contributor I have to *build an emulator* to run the test suite (especially to bless tests that fail in PR CI or full CI) for riscv specific tests or revisions, then that is a very significant burden. For current Tier 1 targets I'm lucky enough to have access to both Windows and Linux via dev-desktop and I don't need to build an emulator for those targets. Even apple-specific failures can already be a pain. I don't want to have to litter `//@ ignore-riscv` into our test suites more than there already exists if there are test failures that are blocking PRs but neither the PR author nor reviewer can bless the test easily.

I have to also note that many of our tests are `//@ ignore-cross-compile` and are not exercised on the cross-compiled target when running tests for various reasons. This means that if a Tier 1 target is emulated it will likely receive a lot less test coverage than you might expect for a Tier 1 target.

Reply:
> Having to set up QEMU, etc. at all in order to debug a contribution failing on RISCV will be a significant and novel additional burden for many contributors

I think of apple OSes as a far worse target than that. There isn't even emulation available. For windows microsoft offers VMs. For other CPUs there's QEMU at least. Apple is worse than all of that.

Despite that I don't have to care about apple because there are dedicated target maintainers. I'd expect a tier 1 target to get the same support from dedicated maintainers.

Reply:
> I think of apple OSes as a far worse target than that. There isn't even emulation available. For windows microsoft offers VMs. For other CPUs there's QEMU at least. Apple is worse than all of that.

True, you can't (legally) emulate the Apple targets. On the other hand, the user base developing for an Apple platform is huge enough that, even if you (the contributor) don't have access it is easy to find a collaborator knowing how a Mac or iPhone work to help. Same story for Windows and Linux. The point is not how easy to *emulate*/*simulate* the target, but how easy for an average contributor to understand the target-specific issues.

(Personally, without checking the maintenance status, I find it surprising that RISC-V can be promoted before WASM.)

Reply:
Yes, what I was eliding here is that e.g. [common RISCV hardware doesn't work with common Linux distros](https://github.com/rust-lang/rust/issues/127180) because of insufficient upstreaming of necessary kernel patches. So even if a contributor DOES get a RISCV device and sets it up, they'll have trouble running rustc on it.

Reply:
> I am not so sure.
> 
> Everyone will be expected to be able to easily contribute to maintaining RISCV targets, because all PRs are blocked on a failure for a tier 1 target's tests. Having to set up QEMU, etc. at all in order to debug a contribution failing on RISCV will be a _significant and novel_ **additional** burden for many contributors, above and beyond what many will expect for contributions, and may easily block PRs. The reason the tier 1 targets that exist now are acceptable (and there is argument that some are rapidly becoming not!) is because it is relatively easy to obtain access to these machines by sheer dint of their commonality, and they usually are okay at performance. The entire reason this proposal exists in its current form is because this target, however, does not fulfill either criterion.
> 
> Yes, it is "merely software", but in this case, running the test suite has to be made as turnkey as possible: none of the existing test infra that runs in CI should be taken as "good enough". No other tier 1 target will have to **always** be emulated in order to effectively run its test suite.
> 
> Fortunately, [there has been work on making testing easier for e.g. better testing the wasm targets](https://github.com/rust-lang/rust/pull/122036), by supporting a "runtool" in our test infrastructure. This makes it possible to simply use [`x.py test --target wasm32-wasip1`](https://github.com/rust-lang/rust/blob/56e35a5dbb37898433a43133dff0398f46d577b8/src/doc/rustc/src/platform-support/wasm32-wasip1.md?plain=1#L122) if one sets the `WASI_SDK_PATH`. So, it should be relatively easy to make testing this target from a different machine host about as simple as `./x.py test --target riscv64gc-unknown-linux-gnu`, and I would expect that **before** this is promoted to tier 1.

You make very good points, thanks. 

We will consult the Rust infrastructure team to try our best to mitigate any significant additional burden. If the net outcome is that this is in fact a significant and unwelcome additional burden then we will seek a path of least resistance as deemed appropriate.

The key point here is that a Tier-1 'badge' significantly improves the quality assurance of Rust for this target, albeit for the qemu-system-riscv64 virt machine 'edition'. 

Given that that very target is the focus of a significant majority of multiple market vertical specific RISC-V enablement, it is not unreasonable to call out the risk (no pun intended) of the impedance introduced in Rust's take up in those domains if we don't move towards a reasonable attainment of that badge.

Once again - that does not mean we want to be cavalier about the intent. If the net outcome of this exercise is an improvement to the process, that's still a win. If we can build enough mindshare to consider a reasonable Tier-1 attainment that would be grand.

All and every opinion is welcome.

Reply:
> > In general, it should be uncomplicated for contributors to build for and use a `riscv64gc-unknown-linux-gnu` emulator like `qemu`, `docker`, or `lima`.
> 
> +1 to what Jubilee said. If as a compiler contributor I have to _build an emulator_ to run the test suite (especially to bless tests that fail in PR CI or full CI) for riscv specific tests or revisions, then that is a very significant burden. For current Tier 1 targets I'm lucky enough to have access to both Windows and Linux via dev-desktop and I don't need to build an emulator for those targets. Even apple-specific failures can already be a pain. I don't want to have to litter `//@ ignore-riscv` into our test suites more than there already exists if there are test failures that are blocking PRs but neither the PR author nor reviewer can bless the test easily.
> 
> I have to also note that many of our tests are `//@ ignore-cross-compile` and are not exercised on the cross-compiled target when running tests for various reasons. This means that if a Tier 1 target is emulated it will likely receive a lot less test coverage than you might expect for a Tier 1 target.

Good points, all.

In the case it is useful and from my personal experience: Every major Linux distro ([debian](https://packages.debian.org/search?searchon=contents&keywords=qemu-system-riscv64&mode=exactfilename&suite=bookworm&arch=any), [ubuntu](https://packages.ubuntu.com/search?searchon=contents&keywords=qemu-system-riscv64&mode=exactfilename&suite=noble&arch=any), [fedora](https://packages.fedoraproject.org/pkgs/qemu/qemu-system-aarch64/), [arch](https://archlinux.org/packages/extra/x86_64/qemu-system-riscv/)),  provides a packaged qemu-system-riscv64 instance. Generally this suffices for running the test suite for the scenario you describe.

Not meaning to condone that as a perfect solution of course but this is what the major os vendors use as a regular port of call at present.

Over a reasonable horizon, it is expected that cloud service providers like [Scaleway](https://labs.scaleway.com/en/em-rv1/) will adopt the best in breed RISC-V systems that then become natural replacements to service such scenarios.

That fact then yields itself to a natural inclination to suggest waiting until that point in time arrives. However, as pointed out previously this ends up hindering the pick-up of Rust in the somewhat unquantifiable period in between.

It is a conundrum, none the less and we are more than willing to explore reasonable ways forward.

Reply:
@raw-bin What is the actual command that you use to run the tests for the emulated target for riscv64gc-unknown-linux-gnu from an x86_64-unknown-linux-gnu host? Have you actually done so? What is the command from an x86_64-pc-windows-msvc host? What about from an aarch64-apple-darwin host? Does it require knowing how to set up and configure QEMU? Does  it require installing other software?

Reply:
The point I am making is that it does not matter how commonly distributed the software, OR hardware for that matter, is, if it is a bitch and a half to actually run the tests.

Reply:
@workingjubilee, we'll get back ASAP with the exact host targets we've used. I would be surprised if the qemu-system-riscv64 invocation from any Tier-1 host would be any different but yes, the fundamental invocation needs to be known. I would assume that that would be reasonably discernible from the CI report itself when a failure occurs and therefore be easily invokable 'locally' but we'll ensure that that gets suitable coverage, either in the RFC text itself or some suitable proxy.

Your point is taken all the same (although speaking for myself I wouldn't use some of the terms you've chosen to make your point but that's just me).

Thanks all the same and please stand by.

Reply:
> riscv64gc-unknown-linux-gnu from an x86_64-unknown-linux-gnu host?

I've run `DEPLOY=1 ./src/ci/docker/run.sh riscv64gc-gnu` (which is exactly what the CI uses) from `aarch64-darwin` and `x86_64-unknown-linux-gnu` several times over the last months. This [uses the `qemu` support](https://github.com/rust-lang/rust/blob/7fc70f870a1d11c3b104274d3a18112996cdec9a/src/ci/docker/host-x86_64/disabled/riscv64gc-gnu/Dockerfile#L94-L97) Rust has already existing.

I've not run the tests from a Windows host.

The experience does leave some things to be desired, for example, it's a bit hard to run a specific test filter using `run.sh`, and there are quite a few container build steps. I think it's fair to say this process is not super obvious and may provide barriers to contributors, I also think this problem is somewhat shared with other targets (such as armhf).

Locally, most of my RISC-V development happens within a [lima](https://github.com/lima-vm/lima/blob/118c249cc8686d65507b9249e931a7feed615497/examples/experimental/riscv64.yaml) VM. All tests pass even using RISC-V host tools here. However this is not the most common tool. It's (almost annoyingly) fairly easy to 'just get' a RISC-V container with the tool.

Running a RISC-V VM in QEMU is generically documented, for example, [here](https://wiki.ubuntu.com/RISC-V/QEMU). Users would need to have (as it documents) `opensbi` and some other usual QEMU dependencies. I believe a user could set `target.riscv64gc-unknown-linux-gnu.qemu-rootfs` in their `config.toml`, and have it work via `x.py`... if they had everything set up correctly. (The problem is: getting there)

I believe if a user has `docker` and `binfmt` configured correctly (sadly, I lack a Linux host with binfmt support) it should be possible to run `docker run --platform linux/riscv64 --rm -ti docker.io/riscv64/ubuntu:24.04` and get a RISC-V container.

It's, unfortunately, not quite the `./x.py test --target "${TARGET}"` we discussed recently.

Reply:
> So to risk a slippery-slope argument, I'm not sure why we would promote RISC-V and not ARM, or potentially many others.

well `riscv64gc-unknown-linux-gnu` does have 4 maintainers listed in https://doc.rust-lang.org/nightly/rustc/platform-support/riscv64gc-unknown-linux-gnu.html proving the "The target maintainer team must include at least 3 developers" requirement, while `arm-unknown-linux-gnueabihf` does not even have a target-specific doc.